File: src/main/java/io/github/suitetecsa/sdk/nauta/scraper/ErrorParser.java
package io.github.suitetecsa.sdk.nauta.scraper;

import io.github.suitetecsa.sdk.nauta.exception.NotLoggedInException;
import io.github.suitetecsa.sdk.nauta.network.HttpResponse;
import io.github.suitetecsa.sdk.nauta.utils.ExceptionHandler;
import io.github.suitetecsa.sdk.nauta.utils.PortalManager;

public interface ErrorParser {
    <T extends Exception> HttpResponse throwExceptionOnFailure(HttpResponse httpResponse, String message, ExceptionHandler<T> exceptionHandler) throws T, NotLoggedInException;

    class Builder {
        private PortalManager portalManager;

        public Builder whenPortalManager(PortalManager portalManager) {
            this.portalManager = portalManager;
            return this;
        }
        public ErrorParser build() {
            return new ErrorParserImpl(portalManager != null ? portalManager : PortalManager.USER);
        }
    }
}

File: src/main/kotlin/io/github/suitetecsa/sdk/nauta/scraper/ErrorParser.kt
package io.github.suitetecsa.sdk.nauta.scraper

import io.github.suitetecsa.sdk.nauta.exception.NotLoggedInException
import io.github.suitetecsa.sdk.nauta.network.HttpResponse
import io.github.suitetecsa.sdk.nauta.utils.ExceptionHandler
import io.github.suitetecsa.sdk.nauta.utils.PortalManager

interface ErrorParser {
    @Throws(NotLoggedInException::class)
    fun <T : Exception> throwExceptionOnFailure(
        httpResponse: HttpResponse,
        message: String,
        exceptionHandler: ExceptionHandler<T>
    ): HttpResponse

    class Builder {
        private var portalManager: PortalManager? = null

        fun whenPortalManager(portalManager: PortalManager): Builder {
            this.portalManager = portalManager
            return this
        }

        fun build(): ErrorParser {
            return ErrorParserImpl(portalManager ?: PortalManager.USER)
        }
    }
}
